### ISSUE TITLE
heatmap(density) from scattered dots?

### ISSUE THREAD

**Issue by xindi-dumbledore** [marcharper/python-ternary#81]
Hi~ I'm wondering whether there is a easy way to plot "density" from pure dots? So I'm able to make a scatter plot from dots but what I would like to have is to get a density heat map. But from my understanding, the current heat map function needs a dictionary feed in.

Thanks a lot!

**Comment by marcharper** [comment]
There's no built in function like matplotlib provides. The easiest way is to write an interpolation function and use `heatmapf` or to compute your own dictionary.

**Comment by fbragman** [comment]
I have tried to compute a 3D histogram based on scatter data of the format [0, 0.8, 0.2], [0.3, 0.4 ,0.3] etc.. I then smooth this histogram with a Gaussian kernel. I finally create a dictionary of the smoothed histogram pointing to each coordinate e.g. (0, 0, 0), (0, 0, 1) ... (1, 1, 1) to be mapped to the simplex space.

[data.txt](https://github.com/marcharper/python-ternary/files/2663639/data.txt)

    xyz = np.loadtxt('data.txt', delimiter=',')
    nbins = 11
    H, b = np.histogramdd((xyz[:, 0], xyz[:, 1], xyz[:, 2]),
                          bins=(nbins, nbins, nbins), range=((0, 1), (0, 1), (0, 1)))
    H = H / np.sum(H)

    # 3D smoothing and interpolation
    from scipy.ndimage.filters import gaussian_filter
    kde = gaussian_filter(H, sigma=2)
    interp_dict = dict()
    binx = np.linspace(0, 1, nbins)
    for i, x in enumerate(binx):
        for j, y in enumerate(binx):
            for k, z in enumerate(binx):
                interp_dict[(i/10, j/10, k/10)] = kde[i, j, k]

    fig, tax = ternary.figure(scale=1)
    tax.heatmap(interp_dict)
    tax.show()

I've attached my result. I'm not entirely too sure what is going on and whether I am calling the `scale` variable correctly. My understanding was that if `scale=1`, `(i, j, k)` had to sum to 1. However, looking at the heatmap example [here](https://github.com/marcharper/python-ternary/blob/master/examples/color_coded_heatmap.py), the `x` that is passed to the heatmap function `f` sum to `1` and not `10`.

Any help would be appreciate @marcharper and thanks again for a beautiful package :)

![result](https://user-images.githubusercontent.com/14344809/49740931-3e438200-fc8d-11e8-961d-25d3e52b1854.jpg)

**Comment by marcharper** [comment]
For heatmaps `i + j + k = scale`. These functions create a partition of the simplex and color each sub-polygon (triangles or hexagons) with the lattice point value. The `scale` parameter is how it defines the lattice.

I think you can simply scale up your points -- in this case there's only one subtriangle (0, 0, 1). Try `scale = 10` and `interp_dict[(i, j, k)] = kde[i, j, k]`.

**Comment by fbragman** [comment]
Yup! Works perfectly - see below.

Thanks again - your package is really helping my upcoming paper submissions

![result_2](https://user-images.githubusercontent.com/14344809/49796815-a2218580-fd35-11e8-9d95-ed7bb65fada0.jpg)

**Comment by marcharper** [comment]
Glad to hear it! Let me know when you get published and I'll add your paper to the citations page.

**Comment by ShiveshM** [comment]
From here, is there a way to scale back the labels on the axis so they go between 0 and 1?

**Comment by marcharper** [comment]
Yes, you can simply overwrite the axes using matplotlib's functions for the outer plot (the usual x-axis and y-axis), or the [analogous ternary functions](https://github.com/marcharper/python-ternary/blob/master/ternary/ternary_axes_subplot.py#L345) if you are using the ternary axes markings on the three triangle sides. See the [example here](https://stackoverflow.com/questions/30568516/override-y-axis-tick-labels-without-affecting-the-graph-shape-in-pyplot) for the matplotlib case.

**Comment by fairliereese** [comment]
Hello,

I've successfully gotten this to work on my data but I can't seem to increase the resolution of the triangles (ie make the triangles smaller). My initial thought was to accomplish this by increasing the scale, but this bafflingly resulted in only the middle part of the simplex being colored as a heatmap. Here is my code using the above sample (which results in the same weirdness as it does with my data), and the resultant plot:

```python
xyz = np.loadtxt('data.txt', delimiter=',')
scale = 20
n_bins = scale+1
import numpy as np

t = (xyz[:, 0], xyz[:, 1], xyz[:, 2])

H, b = np.histogramdd(t,
                      bins=(nbins, nbins, nbins), range=((0, 1), (0, 1), (0, 1)))
H = H / np.sum(H)

# 3D smoothing and interpolation
from scipy.ndimage.filters import gaussian_filter
kde = gaussian_filter(H, sigma=scale/5)
interp_dict = dict()
binx = np.linspace(0, 1, nbins)
for i, x in enumerate(binx):
    for j, y in enumerate(binx):
        for k, z in enumerate(binx):
            interp_dict[(i, j, k)] = kde[i, j, k]

fig, tax = ternary.figure(scale=scale)
tax.heatmap(interp_dict)
tax.show()

# checking for null values
df = pd.DataFrame.from_dict(interp_dict, orient='index')
df['i'] = [b[0] for b in df.index.tolist()]
df['j'] = [b[1] for b in df.index.tolist()]
df['k'] = [b[2] for b in df.index.tolist()]
df.rename({0:'val'}, axis=1, inplace=True)
df.loc[df.val.isnull()]
```
<img width="470" alt="Screen Shot 2022-03-04 at 4 51 27 PM" src="https://user-images.githubusercontent.com/21128281/156860392-60569504-1c70-482e-812d-97e808fd8853.png">

The only thing I've really tried (and can even think of that might have caused this) was adjusting `n_bins` to scale accordingly with `scale`, as you can see in the snippet. It's also worth mentioning that other `scale` values that are not 10 result in other weird subsections of the simplex being filled in. 

I also checked my dictionary for tuples that are null, which I know can sometimes result in gray coloration, but there are no data points that look suspicious. Does anyone with more experience / intuition than I know how I might solve this?

**Comment by manuelbaum** [comment]
This was a very useful piece of code, I think it would be great if it would be part of the ternary codebase!

**Comment by nehasrikn** [comment]
+1 to this! Supremely helpful, and definitely something that would be a nice addition if possible!

> This was a very useful piece of code, I think it would be great if it would be part of the ternary codebase!
